js/tootils/src/index.ts

```diff
@@ -6,4 +10,103 @@ export function wait(n: number) {
 	return new Promise((r) => setTimeout(r, n));
 }
 
+const worker_metadata = new Map<
+	number,
+	{
+		worker_index: number;
+		file: string;
+		process: ChildProcess;
+	}
+>();
+
+import { join } from "path";
+
+const demo_path = join(__dirname, "..", "..", "..", "demo");
+
+function spawn_gradio_app(
+	app: string,
+	workerIndex: number
+): Promise<ChildProcess> {
+	return new Promise((res, rej) => {
+		const _process = spawn(`python`, [`${demo_path}/${app}/run.py`], {
+			shell: true,
+			stdio: "pipe",
+			env: {
+				...process.env,
+				GRADIO_SERVER_PORT: `787${workerIndex}`,
+				PYTHONUNBUFFERED: "true"
+			}
+		});
+		_process.stdout.setEncoding("utf8");
+
+		_process.stdout.on("data", (data) => {
+			const _data = data.toString();
+			const PORT_RE = new RegExp(`:787${workerIndex}`);
```

Comment:
Just FYI this will not work if there are more than 10 tests.

Reply:
What if we create one app that has all the gradio apps as subpaths (via `mount_gradio_app`)? That way we can have more than 10 apps and we can access them as `locahost:7860/app_{i}` for i in [0, k)

Reply:
👍 It would be nicer as we wouldn't have to hardcode the port either or worry about closing it

Reply:
Also a good way to test `mount_gradio_app`. Sometimes that has broken on release unexpectedly.

Reply:
Workers get reused as it is a worker pool, so I guess a better way of phrasing is that tests are tied to workers rather than workers being tied to tests. So it is unlikely to cause issues unless you have more than 20 physical cores. 

That said it is a good point as some contributors may have 24+ core machines. 

I like the sub path idea, I'll try that.